master/buildbot/test/unit/test_reporters_pushover.py

```diff
@@ -0,0 +1,386 @@
+                       num_lines=7),
+            fakedb.LogChunk(logid=60, first_line=0, last_line=1, compressed=0,
+                            content=u'Unicode log with non-ascii (\u00E5\u00E4\u00F6).'),
+            fakedb.Patch(id=99, patch_base64='aGVsbG8sIHdvcmxk',
+                         patch_author='him@foo', patch_comment='foo', subdir='/foo',
+                         patchlevel=3),
+        ])
+        for _id in (20,):
+            self.db.insertTestData([
+                fakedb.BuildProperty(
+                    buildid=_id, name="workername", value="wrk"),
+                fakedb.BuildProperty(
+                    buildid=_id, name="reason", value="because"),
+                fakedb.BuildProperty(
+                    buildid=_id, name="scheduler", value="checkin"),
+                fakedb.BuildProperty(
+                    buildid=_id, name="branch", value="master"),
+            ])
+        res = yield utils.getDetailsForBuildset(self.master, 98, wantProperties=True,
+                                                wantPreviousBuild=wantPreviousBuild)
+        builds = res['builds']
+        buildset = res['buildset']
+
+        @defer.inlineCallbacks
+        def getChangesForBuild(buildid):
+            assert buildid == 20
+            ch = yield self.master.db.changes.getChange(13)
+            defer.returnValue([ch])
+
+        self.master.db.changes.getChangesForBuild = getChangesForBuild
+        defer.returnValue((buildset, builds))
+
+    @defer.inlineCallbacks
+    def setupPushoverNotifier(self, *args, **kwargs):
+        pn = PushoverNotifier(*args, **kwargs)
+        yield pn.setServiceParent(self.master)
+        yield pn.startService()
+        defer.returnValue(pn)
+
+    @defer.inlineCallbacks
+    def test_SendNotification(self):
+        pn = yield self.setupPushoverNotifier(user_key="1234", api_token="abcd",
+                                              otherParams={'sound': "silent"})
+        self._http.expect("post", "/1/messages.json",
+                          params={'user': "1234", 'token': "abcd",
+                                  'sound': "silent", 'message': "Test"},
+                          content_json={'status': 1, 'request': '98765'})
+        n = yield pn.sendNotification({'message': "Test"})
+        j = yield n.json()
+        self.assertEqual(j['status'], 1)
+        self.assertEqual(j['request'], '98765')
+
+    def test_init_enforces_tags_and_builders_are_mutually_exclusive(self):
+        self.assertRaises(config.ConfigErrors,
+                          PushoverNotifier, '1234', 'abcd',
+                          tags=['fast', 'slow'], builders=['a', 'b'])
+
+    def test_init_warns_notifier_mode_all_in_iter(self):
+        self.assertRaisesConfigError(
+            "mode 'all' is not valid in an iterator and must be passed in as a separate string",
+            lambda: PushoverNotifier('1234', 'abcd', mode=['all']))
+
+    @defer.inlineCallbacks
```

Comment:
I think only the above tests are needed. below are just the same, and should be factorized to a test for the baseclass.

you would just subclass and implement sendMessage in a UnderTestNotifier, and then keep the below tests there.